compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -4622,6 +4622,19 @@ public boolean isUnionOfSimpleBasicTypes(BType type) {
         return isSimpleBasicType(type.tag);
     }
 
+    public boolean isUnionOfSimpleBasicTypesExceptNilType(BType type) {
+        if (type.tag == TypeTags.UNION) {
+            Set<BType> memberTypes = ((BUnionType) type).getMemberTypes();
+            for (BType memType : memberTypes) {
+                if (memType.tag == TypeTags.NIL || !isSimpleBasicType(memType.tag)) {
```

Comment:
Shall we add a negative test where the RHS evaluates to true here?

e.g., `int[]|string[]`.